src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs

```diff
@@ -0,0 +1,322 @@
+                                                        pattern Top_, mkAnd,
+                                                        mkBottom, mkCeil,
+                                                        mkEquals, mkIff,
+                                                        mkImplies, mkNot, mkOr,
+                                                        mkTop)
+import           Data.Kore.IndexedModule.MetadataTools (MetadataTools)
+
+{--| 'PredicateProof' is a placeholder for a proof showing that a Predicate
+evaluation was correct.
+--}
+data PredicateProof level = PredicateProof
+    deriving (Show, Eq)
+
+{--| 'GenericPredicate' is a wrapper for predicates used for type safety.
+Should not be exported, and should be treated as an opaque entity which
+can be manipulateds only by functions in this module.
+--}
+newtype GenericPredicate pat = GenericPredicate pat
+    deriving (Show, Eq, Functor, Traversable, Foldable)
+
+{--| 'Predicate' is a user-visible representation for predicates.
+--}
+type Predicate level var = GenericPredicate (PureMLPattern level var)
+
+{--| 'CommonPredicate' follows the generic convention of particularizing types
+to Variable.
+--}
+type CommonPredicate level = Predicate level Variable
+
+{- 'compactPredicatePredicate' removes one level of 'GenericPredicate' which
+sometimes occurs when, say, using Predicates as Traversable.
+-}
+compactPredicatePredicate
+    :: GenericPredicate (GenericPredicate a) -> GenericPredicate a
+compactPredicatePredicate (GenericPredicate x) = x
+
+{- 'stringFromPredicate' extracts a string from a GenericPredicate,
+useful in tests. This could be replaced by a generic extractor, but, for now,
+treating it as an opaque entity seems useful.
+-}
+stringFromPredicate :: GenericPredicate String -> String
+stringFromPredicate (GenericPredicate x) = x
+
+{- 'variableSetFromPredicate' extracts a set of variables from a
+GenericPredicate, useful in tests. This could be replaced by a generic
+extractor, but, for now, treating it as an opaque entity seems useful.
+-}
+variableSetFromPredicate
+    :: GenericPredicate (Set.Set (variable level)) -> Set.Set (variable level)
+variableSetFromPredicate (GenericPredicate vars) = vars
+
+{- 'wrapPredicate' wraps a pattern in a GenericPredicate. This is intended for
+predicate evaluation and tests and should not be used outside of that.
+
+We should consider deleting this and implementing the functionality otherwise.
+-}
+wrapPredicate :: PureMLPattern level var -> Predicate level var
+wrapPredicate patt =
+    if isPredicate patt
+    then GenericPredicate patt
+    else error "Trying to wrap non-predicate as predicate!"
+
+{- 'unwrapPredicate' wraps a pattern in a GenericPredicate. This should be
+not be used outside of that.
+
+We should consider deleting this and implementing the functionality otherwise.
+-}
+unwrapPredicate :: Predicate level var -> PureMLPattern level var
+unwrapPredicate (GenericPredicate p) = p
```

Comment:
Should we have an `impl` module which contains all these relevant for testing and only export the smaller / relevant API from the non-impl module?

Reply:
This is also used in the evaluation package.